File: src/manager/src/main/java/ch/unisg/edpo/manager/KafkaConfig.java
package ch.unisg.edpo.manager;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.autoconfigure.kafka.KafkaProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.kafka.annotation.EnableKafka;
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory;
import org.springframework.kafka.core.*;
import org.springframework.kafka.listener.KafkaListenerErrorHandler;

import java.util.ArrayList;

@Configuration
@EnableKafka
class KafkaConfig {
    private final Logger logger = LoggerFactory.getLogger(KafkaConfig.class);
    private final KafkaProperties kafkaProperties;

    public KafkaConfig(KafkaProperties kafkaProperties) {
        this.kafkaProperties = kafkaProperties;
    }

    @Bean
    public ConsumerFactory<String, byte[]> kafkaConsumerFactory() {
        return new DefaultKafkaConsumerFactory<>(kafkaProperties.buildConsumerProperties());
    }

    @Bean
    public ConcurrentKafkaListenerContainerFactory<String, byte[]> kafkaListenerFactory() {
        var containerFactory = new ConcurrentKafkaListenerContainerFactory<String, byte[]>();
        containerFactory.setConsumerFactory(kafkaConsumerFactory());
        containerFactory.setBatchListener(false);
        return containerFactory;
    }

    @Bean
    public KafkaListenerErrorHandler listenerErrorHandler() {
        return (message, exception) -> {
            logger.error("Error handling message: $message", exception);
            return new Object();
        };
    }

    @Bean
    public ProducerFactory<String, byte[]> kafkaProducerFactory() {
        return new DefaultKafkaProducerFactory<>(kafkaProperties.buildProducerProperties());
    }

    @Bean
    public KafkaTemplate<String, byte[]> kafkaTemplate() {
        return new KafkaTemplate<>(kafkaProducerFactory());
    }
}

File: src/conveyor/src/main/kotlin/ch/unisg/edpo/conveyor/KafkaConfig.kt
package ch.unisg.edpo.conveyor

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.boot.autoconfigure.kafka.KafkaProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.kafka.annotation.EnableKafka
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory
import org.springframework.kafka.core.*
import org.springframework.kafka.listener.KafkaListenerErrorHandler


@Configuration
@EnableKafka
class KafkaConfig(private val kafkaProperties: KafkaProperties) {
    val logger: Logger = LoggerFactory.getLogger(this.javaClass)

    @Bean
    fun kafkaConsumerFactory(): ConsumerFactory<String, ByteArray> =
        DefaultKafkaConsumerFactory(kafkaProperties.buildConsumerProperties())

    @Bean
    fun kafkaListenerFactory(): ConcurrentKafkaListenerContainerFactory<String, ByteArray> {
        val containerFactory = ConcurrentKafkaListenerContainerFactory<String, ByteArray>()
        containerFactory.consumerFactory = kafkaConsumerFactory()
        containerFactory.isBatchListener = false
        return containerFactory
    }

    @Bean
    fun listenerErrorHandler(): KafkaListenerErrorHandler = KafkaListenerErrorHandler { message, exception ->
        logger.error("Error handling message: $message", exception)
    }

    @Bean
    fun kafkaProducerFactory(): ProducerFactory<String, ByteArray> =
        DefaultKafkaProducerFactory(kafkaProperties.buildProducerProperties())

    @Bean
    fun kafkaTemplate() = KafkaTemplate(kafkaProducerFactory())
}
